src/test/shared/clients/mockClients.ts

```diff
@@ -98,3 +99,63 @@ export class MockStsClient implements StsClient {
         this.getCallerIdentity = getCallerIdentity
     }
 }
+
+export class MockS3Client implements S3Client {
+    public constructor (
+        private readonly bucketToRegion: Map<string, string>,
+        private readonly throwError: boolean = false
+    ) { }
+
+    public async getBucketLocation(bucket: string): Promise<S3.GetBucketLocationOutput> {
+        shouldThrowError(this.throwError)
+        const region = this.bucketToRegion.get(bucket)
+        if (!region) {
+            throw new AWSError('bucket not found')
+        }
+
+        const valueToReturn: S3.GetBucketLocationOutput = {
+            LocationConstraint: region
+        }
+
+        return await latencyGenerator(valueToReturn)
+    }
+
+    public async listBuckets(): Promise<S3.ListBucketsOutput> {
+        shouldThrowError(this.throwError)
+        const response: S3.ListBucketsOutput = {
+            Buckets: [],
+            // this is sample text from the docs:
+            // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#listBuckets-property
+            Owner: {
+                DisplayName: 'own-display-name',
+                ID: 'examplee7a2f25102679df27bb0ae12b3f85be6f290b936c4393484be31'
+            }
+        }
+
+        const buckets = this.bucketToRegion.keys()
+        for (const bucket of buckets) {
+            response.Buckets!.push({
+                CreationDate: new Date(),
+                Name: bucket
+            })
+        }
+
+        return await latencyGenerator(response)
+    }
+}
+
+// generates a random call latency from 0-50 ms
+async function latencyGenerator<T>(valueToReturn: T): Promise<T> {
```

Comment:
1. What's the advantage of using a random interval?
2. If we do use a random interval, suggest to log it for reproducability.

Reply:
Random interval is for the `await Promise.all()`

This ensures that everything gets added to the array no matter what. This was more of a concern on an older version that was creating arrays in a map if they didn't exist already.

Reply:
Following up on this, what if I add an override timeout value to be used instead of the random interval?